### PR TITLE
Unit Testing Indexers

### PR DESCRIPTION
**Pull request by shashank-boyapally**
### Description
Unit Tested Indexers Package with a 83% overall coverage. Individual coverages are as below.
factory.go: 100%
local.go: 100%
elastic.go: 78.7%
opensearch.go: 76.6%

### Testing
Test cases for each <filename>.go file can be found in the <filename_test>.go. Each test case covers some part of the code. An example test case is shown below:

```
{"Test 2",
	IndexerConfig{Type: "elastic",
		Servers:            []string{},
		Index:              "go-commons",
		InsecureSkipVerify: false,
	},
	"",
	true,
	httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(http.StatusBadRequest)
	})),
},
```

Along with this example case, many other test cases were included to cover the code. In order to test the .go files the following commands can be followed by the reviewers. Commands to obtain the coverage report is also included.

```
cd indexers 
go get golang.org/x/tools/cmd/cover
go get github.com/onsi/ginkgo/v2
go install -mod=mod github.com/onsi/ginkgo/v2/ginkgo@latest
ginkgo -v -cover --coverprofile cover.out
go tool cover -html=cover.out
```

### Test suite run output
```
[sboyapal@sboyapal indexers]$ ginkgo -v -cover --coverprofile cover.out
Running Suite: Indexers Suite - /home/sboyapal/Documents/forkedCommons/go-commons/indexers
==========================================================================================
Random Seed: 1687959489

Will run 22 of 22 specs
------------------------------
[BeforeSuite] 
/home/sboyapal/Documents/forkedCommons/go-commons/indexers/test_assets_test.go:10
[BeforeSuite] PASSED [0.000 seconds]
------------------------------
Tests for local.go Default behavior of local.go, new() returns err no metrics directory
/home/sboyapal/Documents/forkedCommons/go-commons/indexers/local_test.go:29
• [0.000 seconds]
------------------------------
Tests for local.go Default behavior of local.go, new() returns nil as error
/home/sboyapal/Documents/forkedCommons/go-commons/indexers/local_test.go:34
• [0.000 seconds]
------------------------------
Tests for local.go Default behaviour of local.go, Index() No err is returned
/home/sboyapal/Documents/forkedCommons/go-commons/indexers/local_test.go:70
• [0.000 seconds]
------------------------------
Tests for local.go Default behaviour of local.go, Index() No err is returned
/home/sboyapal/Documents/forkedCommons/go-commons/indexers/local_test.go:76
• [0.000 seconds]
------------------------------
Tests for local.go Default behaviour of local.go, Index() Err is returned metricsdirectory has fault
/home/sboyapal/Documents/forkedCommons/go-commons/indexers/local_test.go:83
• [0.000 seconds]
------------------------------
Tests for local.go Default behaviour of local.go, Index() Err is returned by documents not processed
/home/sboyapal/Documents/forkedCommons/go-commons/indexers/local_test.go:91
• [0.000 seconds]
------------------------------
Factory.go Unit Tests: NewIndexer() Default behaviour of NewIndexer() returns indexer and nil
/home/sboyapal/Documents/forkedCommons/go-commons/indexers/factory_test.go:30
• [0.001 seconds]
------------------------------
Factory.go Unit Tests: NewIndexer() Default behaviour of NewIndexer() returns indexer and err status bad gateway
/home/sboyapal/Documents/forkedCommons/go-commons/indexers/factory_test.go:37
• [0.000 seconds]
------------------------------
Factory.go Unit Tests: NewIndexer() Default behaviour of NewIndexer() returns indexer and err unknown indexer
/home/sboyapal/Documents/forkedCommons/go-commons/indexers/factory_test.go:48
• [0.000 seconds]
------------------------------
Tests for opensearch.go Tests for new() Returns error
/home/sboyapal/Documents/forkedCommons/go-commons/indexers/opensearch_test.go:32
• [0.000 seconds]
------------------------------
Tests for opensearch.go Tests for new() when no url is passed
/home/sboyapal/Documents/forkedCommons/go-commons/indexers/opensearch_test.go:42
• [0.007 seconds]
------------------------------
Tests for opensearch.go Tests for new() Returns err not a valid URL in env variable
/home/sboyapal/Documents/forkedCommons/go-commons/indexers/opensearch_test.go:50
• [0.000 seconds]
------------------------------
Tests for opensearch.go Tests for new() Returns err no index name
/home/sboyapal/Documents/forkedCommons/go-commons/indexers/opensearch_test.go:59
• [0.000 seconds]
------------------------------
Tests for opensearch.go Tests for Index() No err returned
/home/sboyapal/Documents/forkedCommons/go-commons/indexers/opensearch_test.go:98
• [0.044 seconds]
------------------------------
Tests for opensearch.go Tests for Index() err returned docs not processed
/home/sboyapal/Documents/forkedCommons/go-commons/indexers/opensearch_test.go:103
• [0.001 seconds]
------------------------------
Factory.go Unit Tests: NewIndexer() Default behaviour of NewIndexer() returns indexer and nil
/home/sboyapal/Documents/forkedCommons/go-commons/indexers/factory_test.go:77
• [0.001 seconds]
------------------------------
Tests for elastic.go Tests for new() Returns error status bad request
/home/sboyapal/Documents/forkedCommons/go-commons/indexers/elastic_test.go:33
• [0.000 seconds]
------------------------------
Tests for elastic.go Tests for new() when no url is passed
/home/sboyapal/Documents/forkedCommons/go-commons/indexers/elastic_test.go:43
• [0.006 seconds]
------------------------------
Tests for elastic.go Tests for new() Returns err not passing a valid URL in env variable
/home/sboyapal/Documents/forkedCommons/go-commons/indexers/elastic_test.go:52
• [0.000 seconds]
------------------------------
Tests for elastic.go Tests for new() Returns err no index name
/home/sboyapal/Documents/forkedCommons/go-commons/indexers/elastic_test.go:61
• [0.000 seconds]
------------------------------
Tests for elastic.go Tests for Index() No err returned
/home/sboyapal/Documents/forkedCommons/go-commons/indexers/elastic_test.go:101
• [0.041 seconds]
------------------------------
Tests for elastic.go Tests for Index() err returned docs not processed
/home/sboyapal/Documents/forkedCommons/go-commons/indexers/elastic_test.go:106
• [0.001 seconds]
------------------------------

Ran 22 of 22 Specs in 0.104 seconds
SUCCESS! -- 22 Passed | 0 Failed | 0 Pending | 0 Skipped
PASS
        github.com/cloud-bulldozer/go-commons/indexers  coverage: 82.8% of statements
composite coverage: 82.8% of statements

Ginkgo ran 1 suite in 624.890917ms
Test Suite Passed

```

### CODE REVIEW

**Comment by vishnuchalla** [comment]
@shashank-boyapally Thanks for this addition. Here are some overall comments.
* Please integrate a code coverage bot to provide the coverage details.
* The current testing seem to be missing some of the standard unit-test expectations in GO. So requesting to explore and leverage the below libraries.
      - https://github.com/onsi/ginkgo
      - https://github.com/onsi/gomega
      - https://github.com/golang/mock

File: indexers/elastic_test.go [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+			defer testcase.mockServer.Close()
+			testcase.indexerConfig.Servers = []string{testcase.mockServer.URL}
+			testcase.indexerConfig.Index = ""
+			err := indexer.new(testcase.indexerConfig)
+			Expect(err).NotTo(BeNil())
+		})
+
+	})
+
+	Context("Tests for Index()", func() {
```
Comment (vishnuchalla) [inline indexers/elastic_test.go:90]: For indexing I think we can have a success case by re-index this doc every time which should be ok as it keeps on updating the same doc everytime.[Same applies for opensearch.go as well]
```
{
  "_index": "dummy",
  "_id": "Vdc8g4IBYDIaJD7xG1yv",
  "_version": 1,
  "_score": null,
  "_source": {
    "timestamp": "2022-08-09T15:32:10.784425",
    "uuid": "de4f28de-93de-4ab1-90dc-bdd67653b895",
    "connection_time": 0.00347347604110837
  },
  "fields": {
    "timestamp": [
      "2022-08-09T15:32:10.784Z"
    ]
  },
  "sort": [
    1660059130784
  ]
}
```

**Review by vishnuchalla — CHANGES_REQUESTED** [review]
Please update the description with new ways to trigger test and generate coverage report using `ginkgo` framework.

File: indexers/elastic_test.go [hunk 2]
```diff
@@ -0,0 +1,132 @@
+package indexers
+
+import (
+	"net/http"
+	"net/http/httptest"
+	"os"
+
+	. "github.com/onsi/ginkgo"
```
Comment (vishnuchalla) [inline indexers/elastic_test.go:8]: Can we change this to latest: "github.com/onsi/ginkgo/v2"?

**Review by vishnuchalla — APPROVED** [review]
Awesome work. LGTM!

**Comment by vishnuchalla** [comment]
@jtaleric , @rsevilla87 , @chentex Any thoughts on these changes?

File: go.mod [hunk 3]
```diff
@@ -13,14 +13,23 @@ require (
 	k8s.io/utils v0.0.0-20230406110748-d93618cff8a2
 )
 
+require (
+	github.com/fsnotify/fsnotify v1.4.9 // indirect
+	github.com/go-task/slim-sprig v0.0.0-20230315185526-52ccab3ef572 // indirect
+	github.com/google/pprof v0.0.0-20210720184732-4bb14d4b1be1 // indirect
```
Comment (jtaleric) [inline go.mod:19]: This appears out of place?

**Comment by shashank-boyapally** [comment]
removed and tidied go.mod

File: go.mod (hunk 3 above)
Comment (vishnuchalla) [reply go.mod:18]: @shashank-boyapally can you please remove these unused packages?

Comment (shashank-boyapally) [reply go.mod:18]: removed fsnotify.
slim-sprig and pprof are dependencies of ginkgo
Screenshot from 2023-06-28 09-47-18

**Review by jtaleric — APPROVED** [review]
lgtm

**Pull request merged**